controller/backing_image_controller.go

```diff
@@ -0,0 +1,621 @@
+		return err
+	}
+
+	if err := bic.updateDiskLastReferenceMap(backingImage); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getDownloadPodName(biName, diskUUID string) string {
+	return fmt.Sprintf("%s-download-%s", biName, diskUUID[:8])
+}
+
+func (bic *BackingImageController) syncBackingImageWithDownloadPods(backingImage *longhorn.BackingImage) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to sync backing image with download pods")
+	}()
+	log := getLoggerForBackingImage(bic.logger, backingImage)
+
+	if backingImage.Status.DiskDownloadStateMap == nil {
+		backingImage.Status.DiskDownloadStateMap = map[string]types.BackingImageDownloadState{}
+	}
+
+	for diskUUID := range backingImage.Spec.Disks {
+		downloadPodName := getDownloadPodName(backingImage.Name, diskUUID)
+		downloadPod, err := bic.ds.GetPod(downloadPodName)
+		if err != nil {
+			return err
+		}
+		if downloadPod == nil {
+			if _, exists := backingImage.Status.DiskDownloadStateMap[diskUUID]; exists {
+				log.WithField("diskUUID", diskUUID).Warnf("The disk download state becomes %v due to the nil pod",
+					types.BackingImageDownloadStateFailed)
+				backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
+			}
+			// 	else this is a new download request for the disk
+		} else if downloadPod.DeletionTimestamp != nil {
+			log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Warnf("The disk download state becomes %v due to the terminating pod",
+				types.BackingImageDownloadStateFailed)
+			backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
+		} else {
+			switch downloadPod.Status.Phase {
+			case corev1.PodRunning:
+				isReady := false
+				for _, cond := range downloadPod.Status.Conditions {
+					if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+						isReady = true
+					}
+				}
+				if isReady {
+					if backingImage.Status.DiskDownloadStateMap[diskUUID] != types.BackingImageDownloadStateDownloaded {
+						log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Infof("The download state becomes %v from %v since the download pod is ready",
+							types.BackingImageDownloadStateDownloaded, backingImage.Status.DiskDownloadStateMap[diskUUID])
+						backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateDownloaded
+					}
+				} else {
+					if backingImage.Status.DiskDownloadStateMap[diskUUID] == types.BackingImageDownloadStateDownloaded {
+						log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Warnf("The download state becomes %v from %v since the download pod suddenly becomes unavailable",
+							types.BackingImageDownloadStateFailed, types.BackingImageDownloadStateDownloaded)
+						backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
+					}
+				}
```

Comment:
```suggestion
				downloadComplete := backingImage.Status.DiskDownloadStateMap[diskUUID] == types.BackingImageDownloadStateDownloaded
				if !downloadComplete && isReady {
					log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Infof("The download state becomes %v from %v since the download pod is ready",
							types.BackingImageDownloadStateDownloaded, backingImage.Status.DiskDownloadStateMap[diskUUID])
					backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateDownloaded
				} 
				
				if downloadComplete && !isReady {
					log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Warnf("The download state becomes %v from %v since the download pod suddenly becomes unavailable",
							types.BackingImageDownloadStateFailed, types.BackingImageDownloadStateDownloaded)
					backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
				}
```

Add a downloadComplete flag, that will make the control flow logic simpler.
Since this functions control flow structure should be refactored further see above comment this will also reduce the nesting further.